src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs

```diff
@@ -12,48 +12,44 @@
 namespace Microsoft.VisualStudio.LanguageServices.Experimentation
 {
     [Export(typeof(VisualStudioExperimentationService))]
-    [ExportWorkspaceService(typeof(IExperimentationService), ServiceLayer.Host), Shared]
-    internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedObject, IExperimentationService
+    [ExportWorkspaceService(typeof(IExperimentationServiceFactory), ServiceLayer.Host), Shared]
+    internal class VisualStudioExperimentationService : IExperimentationServiceFactory
     {
-        private readonly object _experimentationServiceOpt;
-        private readonly MethodInfo _isCachedFlightEnabledInfo;
-        private readonly IVsFeatureFlags _featureFlags;
+        private readonly IAsyncServiceProvider _serviceProvider;
+        private IExperimentationService _experimentationService;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public VisualStudioExperimentationService(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
-            : base(threadingContext)
+        public VisualStudioExperimentationService(SVsServiceProvider serviceProvider)
         {
-            object experimentationServiceOpt = null;
-            MethodInfo isCachedFlightEnabledInfo = null;
-            IVsFeatureFlags featureFlags = null;
+            _serviceProvider = (IAsyncServiceProvider)serviceProvider;
+        }
 
-            threadingContext.JoinableTaskFactory.Run(async () =>
+        public async ValueTask<IExperimentationService> GetExperimentationServiceAsync(CancellationToken cancellationToken)
+        {
+            if (_experimentationService is null)
             {
-                try
-                {
-                    featureFlags = (IVsFeatureFlags)await ((IAsyncServiceProvider)serviceProvider).GetServiceAsync(typeof(SVsFeatureFlags)).ConfigureAwait(false);
-                    experimentationServiceOpt = await ((IAsyncServiceProvider)serviceProvider).GetServiceAsync(typeof(SVsExperimentationService)).ConfigureAwait(false);
-                    if (experimentationServiceOpt != null)
-                    {
-                        isCachedFlightEnabledInfo = experimentationServiceOpt.GetType().GetMethod(
-                            "IsCachedFlightEnabled", BindingFlags.Public | BindingFlags.Instance);
-                    }
-                }
-                catch
-                {
-                }
-            });
+                var featureFlags = (IVsFeatureFlags)await _serviceProvider.GetServiceAsync(typeof(SVsFeatureFlags)).ConfigureAwait(false);
```

Comment:
This is still a deadlock risk since the IVsFeatureFlags will cause COM RPC and risk deadlock -- are we following a rule that it shouldn't be waited for on the UI thread?

Reply:
This is updated to perform the casts on the main thread until we verify that both interface implementations are managed objects.

Reply:
If they are both managed objects, then it means the only jump to the UI thread behind GetServiceAsync was because the services weren't marked as free-threaded, and maybe this gives a different line of attack -- are the implementations actually free-threaded in practice but just need to be marked as such? If so, we can just ask the shell to fix that, this code then all becomes correct and the PR isn't needed at all.